posting.go

```diff
@@ -583,7 +583,7 @@ func (i *PostingsIterator) nextDocNumAtOrAfter(atOrAfter uint64) (uint64, bool,
 	// n is the next actual hit (excluding some postings), and
 	// allN is the next hit in the full postings, and
 	// if they don't match, move 'all' forwards until they do
-	for allN != n {
+	for allN != n && i.all.HasNext() {
```

Comment:
If you check the code for Next() then it internally has a call for hasNext().  (Thinking if Next() crashes then HasNext() can crash as well).
So, I don't think this explicit check would fix the crash.

Rather, the former PR which exits early upon a "i.postings == nil || i.postings == emptyPostingsList" would fix/improve the situation seems more intuitive.

Reply:
> If you check the code for Next() then it internally has a call for hasNext(). (Thinking if Next() crashes then HasNext() can crash as well).

That's not accurate though. Next() dereferences a pointer before it calls hasNext() (whose signature differs from HasNext()). We always check i.Actual's HasNext() before calling Next() but it isn't the same for i.all.